### PR TITLE
Support CLI commands testing

### PR DESCRIPTION
**Pull request by hidmic**
This pull request encompasses all additions to `launch`, and mainly `launch_testing`, to allow for CLI commands testing. 

Connected to https://github.com/ros2/ros2cli/pull/304.

### CODE REVIEW

File: launch_testing/launch_testing/proc_info_handler.py [hunk 1]
```diff
@@ -100,14 +100,11 @@ def __getitem__(self, key):
...
-                              *,
-                              timeout=10):
-
-        success = False
-
+    def waitForShutdown(self,
+                        process,
+                        cmd_args=None,
+                        *,
+                        timeout=10):
```
Comment (dirk-thomas) [inline launch_testing/launch_testing/proc_info_handler.py]: I see that the code was using alignment before but our style guide explicitly avoids that. Please wrap after the open parenthesis and indent the following lines one level.

Comment (hidmic) [reply launch_testing/launch_testing/proc_info_handler.py]: Good catch. Done in 60f12a0d2144e7dc88ded827f6ac3390e80f7643.

**Comment by hidmic** [comment]
This fixes issues in `launch_testing` and that's regardless of CLI testing. I'll run CI on this patch only so we can get it in.

**Comment by hidmic** [comment]
Same as in https://github.com/ros2/ros2cli/pull/304, https://github.com/ros2/launch/pull/279/commits/9891ee09d5738862e8946d68d6b1d13817d5d5a5 introduces significant changes and improvements.

File: launch_testing/launch_testing/tools/output.py [hunk 2]
```diff
@@ -77,3 +82,49 @@ def expected_output_from_file(path):
...
+    """
+    if (text is not None) == (lines is not None):
+        raise ValueError('Either lines or text, but not both, must be specified')
+
+    if (expected_text is not None) == (expected_lines is not None):
+        raise ValueError('Either expected lines or text, but not both, must be specified')
+
+    if expected_text is not None:
+        if text is None:
+            text = '\n'.join(lines)
```
Comment (ivanpauno) [inline launch_testing/launch_testing/tools/output.py]: IMO, documentation is not clear enough about this.
I thought you could use both `lines` and `expected_lines`, or `text` and `expected_text`. But I didn't get than combinations like `lines` and `expected_text` were valid.

File: launch_testing/launch_testing/tools/output.py [hunk 3]
```diff
@@ -77,3 +82,49 @@ def expected_output_from_file(path):
             return [re.compile(regex) for regex in f.read().splitlines()]
 
     raise RuntimeError('could not find output check file: {}'.format(path))
+
+
+def expect_output(
+    text=None,
```
Comment (ivanpauno) [inline launch_testing/launch_testing/tools/output.py]: Maybe, it's clearer if you only use keyworded arguments.

File: launch_testing/launch_testing/markers.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        def _wrapper(*args, **kwargs):
+            n = times
+            while n > 1:
+                try:
+                    return func(*args, **kwargs)
+                except AssertionError:
+                    n -= 1
+            return func(*args, **kwargs)
+        # Keep signature for argument binding
+        _wrapper.__signature__ = inspect.signature(func)
```
[the remark this answers is not in the document]
Comment (hidmic) [reply launch_testing/launch_testing/markers.py]: > and `functools.wraps` preserves the signature

Not for the `inspect` module. Try [`inspect.getfullargspec()`](https://docs.python.org/3.6/library/inspect.html#inspect.getfullargspec) on a decorated function that has no `__signature__`.

[the remark this answers is not in the document]
Comment (hidmic) [reply launch_testing/launch_testing/markers.py]: After an f2f with @ivanpauno and some additional digging into `inspect`'s documentation, `inspect.signature()` provides an alternative to `inspect.getfullargspec()` but while the former abides to the `__wrapped__` attribute that `functools.wraps()` gives the latter doesn't. `inspect .getfullargspec()` was actually deprecated in favor of `inspect.signature` in Python 3.5 and then brought back in 3.6 (see [here](https://docs.python.org/3.6/library/inspect.html#inspect.getfullargspec)).

Conclusion: we're moving away from `inspect.getfullargspec()`.

File: launch_testing/launch_testing/tools/output.py (hunk 3 above)
Comment (hidmic) [reply launch_testing/launch_testing/tools/output.py]: That was the case originally, but not requiring `text` to be a keyword plays its role when `expect_output` is used as a condition for `ProcessProxy.wait_for_output` (or we force a boilerplate lambda). See [here](https://github.com/ros2/launch/pull/279/files#diff-e6ac5ee5445a9a82cfbe47983f5e4529R66-R84).

File: launch_testing/launch_testing/tools/output.py (hunk 2 above)
Comment (hidmic) [reply launch_testing/launch_testing/tools/output.py]: Done in d3982e54e9449d63d353bf0a00753b3489b75e7d

File: launch_testing/launch_testing/markers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply launch_testing/launch_testing/markers.py]: Done in e007f549299d7c56c5e0ca978af6f28fd7eda257.

**Comment by hidmic** [comment]
CI (full, against Fast-RTPS, Connext and Opensplice)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8398/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4313/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6837/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8297/)

File: launch_testing/launch_testing/tools/output.py (hunk 3 above)
Comment (ivanpauno) [reply launch_testing/launch_testing/tools/output.py]: Makes sense

**Pull request merged**